### PR TITLE
level map, a coefficient coder for av1

### PR DESCRIPTION
**Pull request by ycho**
For #106.

### CODE REVIEW

**Comment by ycho** [comment]
Will need updated libaom, as PR-ed https://github.com/tdaede/aom/pull/10

File: src/ec.rs [hunk 1]
```diff
@@ -277,6 +277,26 @@ impl Writer {
...
+        let mut i = x;
+        let mut length = 0;
+
+        while i != 0 {
+          i >>= 1;
+          length += 1;
+        }
+        assert!(length > 0);
+
+        for _i in 0..length-1 { self.bit(0); };
```
[the remark this answers is not in the document]
Comment (ycho) [reply src/ec.rs]: Thanks for the nice tip! What a rust!

File: src/ec.rs [hunk 2]
```diff
@@ -277,6 +277,26 @@ impl Writer {
...
+        let p = ((0x7FFFFF - (128 << 15) + 128) >> 8) as u16;
+        self.enc.od_ec_encode_bool_q15(bit == 1, p);
+    }
+
+    pub fn write_golomb(&mut self, level: u16) {
+        let x = level + 1;
+        let mut i = x;
+        let mut length = 0;
+
+        while i != 0 {
```
[the remark this answers is not in the document]
Comment (ycho) [reply src/ec.rs]: If you don't mind, can you improve it once this patch has landed?

[the remark this answers is not in the document]
Comment (tdaede) [reply src/ec.rs]: The leading_zeroes / trailing_zeroes are faster as they can use the bsr instruction, but I think we can fix it up afterwards.

File: src/context.rs [hunk 3]
```diff
@@ -46,24 +50,46 @@ pub static tx_size_wide_log2: [usize; TX_SIZES_ALL] =
     [2, 3, 4, 5, 2, 3, 3, 4, 4, 5, 2, 4, 3, 5];
 pub static tx_size_high_log2: [usize; TX_SIZES_ALL] =
     [2, 3, 4, 5, 3, 2, 4, 3, 5, 4, 4, 2, 5, 3];
+// Transform block width in pixels
+pub static tx_size_wide: [usize; TX_SIZES_ALL] =
```
Comment (tdaede) [inline src/context.rs]: I want to add width() and height() to the enum like bsize has, but that can wait.

File: src/context.rs [hunk 4]
```diff
@@ -312,6 +420,374 @@ const REF_TYPES: usize = 2;
...
+    10, 10, 10, 10, 10, 10, 10, 10,  // 257-512
+    11                               // 513-
+];
+
+
+static k_eob_group_start: [u16; 12] = [ 0, 1, 2, 3, 5, 9,
+                                        17, 33, 65, 129, 257, 513 ];
+static k_eob_offset_bits: [u16; 12] = [ 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ];
+
+static clip_max3: [u8; 256] = [
```
Comment (tdaede) [inline src/context.rs]: I know this is probably how libaom does this clip... but I really doubt using a lookup table is faster than a normal clip here.

**Comment by tdaede** [comment]
r+

**Pull request merged**